File: src/layer_shell_py/instance.py
from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum
from os import environ, getuid, kill
from pathlib import Path
from re import fullmatch
from signal import SIGTERM
from socket import AF_UNIX, SOCK_STREAM, socket
from subprocess import DEVNULL, Popen
from sys import argv
from time import monotonic, sleep


class InstanceAction(StrEnum):
    PRELOAD = "preload"
    START = "start"
    STOP = "stop"
    TOGGLE = "toggle"
    QUIT = "quit"


class InstanceError(RuntimeError):
    pass


@dataclass(frozen=True)
class Instance:
    effect: str
    instance_id: str

    @property
    def pid_path(self) -> Path:
        return _runtime_dir() / f"{self.effect}-{self.instance_id}.pid"

    @property
    def log_path(self) -> Path:
        return _runtime_dir() / f"{self.effect}-{self.instance_id}.log"

    @property
    def socket_path(self) -> Path:
        return _runtime_dir() / f"{self.effect}-{self.instance_id}.sock"


def validate_instance_id(instance_id: str) -> str:
    if fullmatch(r"[A-Za-z0-9_.-]+", instance_id) is None:
        msg = (
            "Instance id may only contain letters, numbers, dots, underscores, "
            "and hyphens."
        )
        raise InstanceError(msg)
    return instance_id


def manage_instance(
    *,
    action: InstanceAction,
    effect: str,
    instance_id: str,
    child_args: list[str],
) -> int:
    instance = Instance(effect=effect, instance_id=validate_instance_id(instance_id))

    if action is InstanceAction.PRELOAD:
        return ensure_resident(
            instance,
            child_args,
            initial_visible=False,
            command=None,
        )

    if action is InstanceAction.START:
        return ensure_resident(
            instance,
            child_args,
            initial_visible=True,
            command="show",
        )

    if action is InstanceAction.TOGGLE:
        return ensure_resident(
            instance,
            child_args,
            initial_visible=True,
            command="toggle",
        )

    if action is InstanceAction.STOP:
        return send_command(instance, "hide", missing_ok=True)

    if action is InstanceAction.QUIT:
        return quit_instance(instance)

    msg = f"Unsupported instance action: {action}"
    raise InstanceError(msg)


def ensure_resident(
    instance: Instance,
    child_args: list[str],
    *,
    initial_visible: bool,
    command: str | None,
) -> int:
    if is_instance_running(instance):
        if command is not None:
            return send_command(instance, command, missing_ok=False)
        return 0

    instance.pid_path.parent.mkdir(parents=True, exist_ok=True)
    instance.socket_path.unlink(missing_ok=True)
    log_file = instance.log_path.open("ab")
    resident_args = [
        *child_args,
        "--resident",
        "--socket-path",
        str(instance.socket_path),
    ]
    if initial_visible:
        resident_args.append("--initial-visible")
    process = Popen(
        [argv[0], *resident_args],
        stdin=DEVNULL,
        stdout=DEVNULL,
        stderr=log_file,
        start_new_session=True,
    )
    log_file.close()
    instance.pid_path.write_text(f"{process.pid}\n", encoding="utf-8")
    return 0


def quit_instance(instance: Instance) -> int:
    if not is_instance_running(instance):
        _remove_stale_pid(instance)
        return 0

    if _try_send_command(instance, "quit"):
        instance.pid_path.unlink(missing_ok=True)
        instance.socket_path.unlink(missing_ok=True)
        return 0

    pid = _read_active_pid(instance)
    if pid is None:
        _remove_stale_pid(instance)
        return 0

    kill(pid, SIGTERM)
    instance.pid_path.unlink(missing_ok=True)
    instance.socket_path.unlink(missing_ok=True)
    return 0


def send_command(instance: Instance, command: str, *, missing_ok: bool) -> int:
    if not is_instance_running(instance):
        if missing_ok:
            _remove_stale_pid(instance)
            return 0
        msg = f"No running {instance.effect} instance: {instance.instance_id}"
        raise InstanceError(msg)

    if _try_send_command(instance, command):
        return 0

    instance.socket_path.unlink(missing_ok=True)
    if not is_instance_running(instance):
        _remove_stale_pid(instance)
    if missing_ok:
        return 0

    msg = f"Could not reach {instance.effect} instance: {instance.instance_id}"
    raise InstanceError(msg)


def _try_send_command(instance: Instance, command: str) -> bool:
    deadline = monotonic() + 1.0
    while True:
        try:
            with socket(AF_UNIX, SOCK_STREAM) as client:
                client.connect(str(instance.socket_path))
                client.sendall(f"{command}\n".encode())
            return True
        except OSError:
            if monotonic() >= deadline:
                return False
            sleep(0.01)


def is_instance_running(instance: Instance) -> bool:
    return _read_active_pid(instance) is not None


def _read_active_pid(instance: Instance) -> int | None:
    try:
        pid = int(instance.pid_path.read_text(encoding="utf-8").strip())
    except FileNotFoundError, ValueError:
        return None

    if pid <= 0 or not _pid_exists(pid) or not _pid_looks_managed(pid, instance.effect):
        return None

    return pid


def _pid_exists(pid: int) -> bool:
    try:
        kill(pid, 0)
    except ProcessLookupError:
        return False
    except PermissionError:
        return True
    return True


def _pid_looks_managed(pid: int, effect: str) -> bool:
    cmdline_path = Path(f"/proc/{pid}/cmdline")
    if not cmdline_path.exists():
        return True

    try:
        cmdline = cmdline_path.read_bytes().decode(errors="replace")
    except OSError:
        return False

    return "layer-shell-py" in cmdline and effect in cmdline


def _remove_stale_pid(instance: Instance) -> None:
    instance.pid_path.unlink(missing_ok=True)
    instance.socket_path.unlink(missing_ok=True)


def _runtime_dir() -> Path:
    if xdg_runtime_dir := environ.get("XDG_RUNTIME_DIR"):
        return Path(xdg_runtime_dir) / "layer-shell-py"

    return Path("/tmp") / f"layer-shell-py-{getuid()}"

File: src/instance.rs
use std::env;
use std::fmt::{self, Display};
use std::fs::{self, OpenOptions};
use std::io::{self, Write};
use std::os::unix::net::UnixStream;
use std::os::unix::process::CommandExt;
use std::path::PathBuf;
use std::process::{Command, Stdio};
use std::thread::sleep;
use std::time::{Duration, Instant};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum InstanceAction {
    Preload,
    Show,
    Hide,
    Toggle,
    Quit,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Instance {
    pub effect: String,
    pub instance_id: String,
}

impl Instance {
    pub fn new(effect: impl Into<String>, instance_id: impl Into<String>) -> Self {
        Self {
            effect: effect.into(),
            instance_id: instance_id.into(),
        }
    }

    pub fn pid_path(&self) -> PathBuf {
        runtime_dir().join(format!("{}-{}.pid", self.effect, self.instance_id))
    }

    pub fn log_path(&self) -> PathBuf {
        runtime_dir().join(format!("{}-{}.log", self.effect, self.instance_id))
    }

    pub fn socket_path(&self) -> PathBuf {
        runtime_dir().join(format!("{}-{}.sock", self.effect, self.instance_id))
    }
}

pub fn validate_instance_id(instance_id: &str) -> Result<(), InstanceError> {
    if instance_id.is_empty()
        || !instance_id
            .chars()
            .all(|ch| ch.is_ascii_alphanumeric() || matches!(ch, '.' | '_' | '-'))
    {
        return Err(InstanceError::InvalidInstanceId);
    }

    Ok(())
}

pub fn manage_instance(
    action: InstanceAction,
    effect: &str,
    instance_id: &str,
    child_args: &[String],
) -> Result<i32, InstanceError> {
    validate_instance_id(instance_id)?;
    let instance = Instance::new(effect, instance_id);

    match action {
        InstanceAction::Preload => ensure_resident(&instance, child_args, false, None),
        InstanceAction::Show => ensure_resident(&instance, child_args, true, Some("show")),
        InstanceAction::Toggle => ensure_resident(&instance, child_args, true, Some("toggle")),
        InstanceAction::Hide => send_command(&instance, "hide", true),
        InstanceAction::Quit => quit_instance(&instance),
    }
}

pub fn ensure_resident(
    instance: &Instance,
    child_args: &[String],
    initial_visible: bool,
    command: Option<&str>,
) -> Result<i32, InstanceError> {
    if is_instance_running(instance) {
        if let Some(command) = command {
            return send_command(instance, command, false);
        }
        return Ok(0);
    }

    fs::create_dir_all(
        instance
            .pid_path()
            .parent()
            .expect("runtime paths always have parents"),
    )?;
    let _ = fs::remove_file(instance.socket_path());

    let log_file = OpenOptions::new()
        .create(true)
        .append(true)
        .open(instance.log_path())?;

    let current_exe = env::current_exe()?;
    let mut resident_args = child_args.to_vec();
    resident_args.extend([
        "--resident".to_string(),
        "--socket-path".to_string(),
        instance.socket_path().display().to_string(),
    ]);
    if initial_visible {
        resident_args.push("--initial-visible".to_string());
    }

    let mut child_command = Command::new(current_exe);
    child_command
        .args(resident_args)
        .stdin(Stdio::null())
        .stdout(Stdio::null())
        .stderr(Stdio::from(log_file));

    // Match Python's start_new_session=True so resident instances survive
    // short-lived CLI invocations cleanly.
    unsafe {
        child_command.pre_exec(|| {
            if libc::setsid() == -1 {
                return Err(io::Error::last_os_error());
            }
            Ok(())
        });
    }

    let child = child_command.spawn()?;
    fs::write(instance.pid_path(), format!("{}\n", child.id()))?;
    eprintln!(
        "Started {effect} instance {instance_id} as PID {pid}. Log: {log}. Socket: {socket}",
        effect = instance.effect,
        instance_id = instance.instance_id,
        pid = child.id(),
        log = instance.log_path().display(),
        socket = instance.socket_path().display(),
    );
    Ok(0)
}

pub fn quit_instance(instance: &Instance) -> Result<i32, InstanceError> {
    if !is_instance_running(instance) {
        remove_stale_pid(instance);
        return Ok(0);
    }

    if try_send_command(instance, "quit") {
        remove_stale_pid(instance);
        return Ok(0);
    }

    if let Some(pid) = read_active_pid(instance) {
        unsafe {
            libc::kill(pid, libc::SIGTERM);
        }
    }
    remove_stale_pid(instance);
    Ok(0)
}

pub fn send_command(
    instance: &Instance,
    command: &str,
    missing_ok: bool,
) -> Result<i32, InstanceError> {
    if !is_instance_running(instance) {
        if missing_ok {
            remove_stale_pid(instance);
            return Ok(0);
        }
        return Err(InstanceError::NoRunningInstance {
            effect: instance.effect.clone(),
            instance_id: instance.instance_id.clone(),
        });
    }

    if try_send_command(instance, command) {
        return Ok(0);
    }

    let _ = fs::remove_file(instance.socket_path());
    if !is_instance_running(instance) {
        remove_stale_pid(instance);
    }
    if missing_ok {
        return Ok(0);
    }

    Err(InstanceError::CouldNotReachInstance {
        effect: instance.effect.clone(),
        instance_id: instance.instance_id.clone(),
    })
}

pub fn try_send_command(instance: &Instance, command: &str) -> bool {
    let deadline = Instant::now() + Duration::from_secs(1);

    loop {
        match UnixStream::connect(instance.socket_path()) {
            Ok(mut stream) => {
                let _ = writeln!(stream, "{command}");
                return true;
            }
            Err(_) if Instant::now() < deadline => sleep(Duration::from_millis(10)),
            Err(_) => return false,
        }
    }
}

pub fn is_instance_running(instance: &Instance) -> bool {
    read_active_pid(instance).is_some()
}

fn read_active_pid(instance: &Instance) -> Option<i32> {
    let pid_text = fs::read_to_string(instance.pid_path()).ok()?;
    let pid: i32 = pid_text.trim().parse().ok()?;

    if pid <= 0 || !pid_exists(pid) || !pid_looks_managed(pid, &instance.effect) {
        return None;
    }

    Some(pid)
}

fn pid_exists(pid: i32) -> bool {
    match unsafe { libc::kill(pid, 0) } {
        0 => true,
        _ => io::Error::last_os_error().raw_os_error() == Some(libc::EPERM),
    }
}

fn pid_looks_managed(pid: i32, effect: &str) -> bool {
    let cmdline_path = PathBuf::from(format!("/proc/{pid}/cmdline"));
    if !cmdline_path.exists() {
        return true;
    }

    let Ok(cmdline) = fs::read_to_string(cmdline_path) else {
        return false;
    };

    cmdline.contains("layer-shell-rs") && cmdline.contains(effect)
}

fn remove_stale_pid(instance: &Instance) {
    let _ = fs::remove_file(instance.pid_path());
    let _ = fs::remove_file(instance.socket_path());
}

pub fn runtime_dir() -> PathBuf {
    if let Some(xdg_runtime_dir) = env::var_os("XDG_RUNTIME_DIR") {
        return PathBuf::from(xdg_runtime_dir).join("layer-shell-rs");
    }

    PathBuf::from("/tmp").join(format!("layer-shell-rs-{}", unsafe { libc::getuid() }))
}

#[derive(Debug)]
pub enum InstanceError {
    InvalidInstanceId,
    NoRunningInstance { effect: String, instance_id: String },
    CouldNotReachInstance { effect: String, instance_id: String },
    Io(io::Error),
}

impl Display for InstanceError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::InvalidInstanceId => formatter.write_str(
                "Instance id may only contain letters, numbers, dots, underscores, and hyphens.",
            ),
            Self::NoRunningInstance {
                effect,
                instance_id,
            } => {
                write!(formatter, "No running {effect} instance: {instance_id}")
            }
            Self::CouldNotReachInstance {
                effect,
                instance_id,
            } => {
                write!(
                    formatter,
                    "Could not reach {effect} instance: {instance_id}"
                )
            }
            Self::Io(error) => Display::fmt(error, formatter),
        }
    }
}

impl std::error::Error for InstanceError {}

impl From<io::Error> for InstanceError {
    fn from(error: io::Error) -> Self {
        Self::Io(error)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn validate_instance_id_accepts_safe_ids() {
        validate_instance_id("fuzzel_1.2-3").unwrap();
    }

    #[test]
    fn validate_instance_id_rejects_paths() {
        assert_eq!(
            validate_instance_id("bad/id").unwrap_err().to_string(),
            "Instance id may only contain letters, numbers, dots, underscores, and hyphens."
        );
    }
}
